node/src/Pos/Explorer/Txp/Local.hs

```diff
@@ -91,9 +93,14 @@ instance MonadTxExtraRead EProcessTxMode where
         HM.lookup addr . eetAddrBalances <$> view eptcExtraBase
 
 eTxProcessTransaction
-    :: ETxpLocalWorkMode ctx m
-    => (TxId, TxAux) -> ExceptT ToilVerFailure m ()
-eTxProcessTransaction itw@(txId, TxAux {taTx = UnsafeTx {..}}) = do
+    :: (ETxpLocalWorkMode ctx m, HasLens' ctx BlkSemaphore, MonadMask m)
+    => (TxId, TxAux) -> m (Either ToilVerFailure ())
+eTxProcessTransaction itw = withBlkSemaphoreIgnoreTip $ eTxProcessTransactionNoLock itw
+
+eTxProcessTransactionNoLock
```

Comment:
For cases like this, where we have bits of code that assume that the lock is held, we could use a typing trick to make sure that these functions are only called when the lock is indeed held. In this specific case it is obvious, but there are now lots of other places in the code elsewhere that are only safe to call when the lock is held.

The `withBlkSemaphore` could provide the proxy to the inner action, and these actions that need to be called with the block held could require it in their type. It'd be a good sanity check and would enforce the pattern we have in various places of `blah` and `blahNoLock`.

Reply:
Good idea, I like it, but will leave it for the next PR :)